src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs

```diff
@@ -88,48 +88,44 @@ internal void AddEntries(TreeRouteBuilder builder, ActionDescriptorCollection ac
             foreach (var routeInfo in routeInfos)
             {
+                var defaults = new RouteValueDictionary();
+                foreach (var kvp in routeInfo.ActionDescriptor.RouteValues)
+                {
+                    defaults.Add(kvp.Key, kvp.Value);
+                }
+
                 builder.MapOutbound(
-                    _handler,
+                    NullRouter.Instance,
                     routeInfo.RouteTemplate, 
-                    new RouteValueDictionary(routeInfo.ActionDescriptor.RouteValueDefaults),
+                    defaults,
                     routeInfo.RouteName, 
                     routeInfo.Order);
             }
 
             // We're creating one AttributeRouteMatchingEntry per group, so we need to identify the distinct set of
             // groups. It's guaranteed that all members of the group have the same template and precedence,
             // so we only need to hang on to a single instance of the RouteInfo for each group.
-            var distinctRouteInfosByGroup = GroupRouteInfosByGroupId(routeInfos);
-            foreach (var routeInfo in distinctRouteInfosByGroup)
+            var groups = GroupRouteInfos(routeInfos);
+            foreach (var group in groups)
             {
+                var handler = _handlerFactory(group.ToArray());
+
                 // Note that because we only support 'inline' defaults, each routeInfo group also has the same
                 // set of defaults. 
                 //
                 // We then inject the route group as a default for the matcher so it gets passed back to MVC
                 // for use in action selection.
-                var entry = builder.MapInbound(
-                    _handler,
-                    routeInfo.RouteTemplate,
-                    routeInfo.RouteName,
-                    routeInfo.Order);
-
-                entry.Defaults[TreeRouter.RouteGroupKey] = routeInfo.RouteGroup;
+                builder.MapInbound(
+                    handler,
+                    group.Key.RouteTemplate,
+                    group.Key.RouteName,
+                    group.Key.Order);
             }
         }
 
-        private static IEnumerable<RouteInfo> GroupRouteInfosByGroupId(List<RouteInfo> routeInfos)
+        private static IEnumerable<IGrouping<RouteInfo, ActionDescriptor>> GroupRouteInfos(List<RouteInfo> routeInfos)
         {
-            var routeInfosByGroupId = new Dictionary<string, RouteInfo>(StringComparer.OrdinalIgnoreCase);
-
-            foreach (var routeInfo in routeInfos)
-            {
-                if (!routeInfosByGroupId.ContainsKey(routeInfo.RouteGroup))
-                {
-                    routeInfosByGroupId.Add(routeInfo.RouteGroup, routeInfo);
-                }
-            }
-
-            return routeInfosByGroupId.Values;
+            return routeInfos.GroupBy(ri => ri, ri => ri.ActionDescriptor, RouteInfoEqualityComparer.Instance);
```

Comment:
We usually use one lettered (throw away) names or full names. `ri` seems like an abbreviation. Anyway, not super concerned about this.